### ISSUE TITLE
Onboarding widgets cover Editor-Content after additional WorbenchWindow-Close

### ISSUE THREAD

**Issue by Rainer7000** [eclipse-platform/eclipse.platform.ui#1180]
This Problem is a bit tricky to reproduce, because it involves a few preconditions.
The Problem was introduced with Eclipse 2023-09, alle previous versions workt fine.
I used the old EMail-Example-Application in Eclipse3 compatibility as a starting-point to reproduce it.
I have a Toolbar-Action that opens verry simple EditorParts (just a left and a right label) which works fine.
A second Toolbar-Action opens an additional Toolbar-Window with a different Perspective.
If the additional Toolbar-Window is closed while the editorArea of the primary WorkbenchWindow ist empty,
the content of the editor is never rendered (it's not a sizing/layouting problem, i checked that with a ControlListener).
The Problem does not occur with ViewParts instead of EditorParts and it also does not occur when one EditorPart 
is still visible when the additional window ist closed.
![rcp_editors](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/efbaa701-5cdb-4c92-bd73-0756d6e12746)
![rcp_window_open](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/b3623b2d-5a77-41cd-92dd-72da22e5b6a2)
![rcp_editors_empty](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/3bc5b85b-6fcf-4167-9a3f-80f957b39416)

**Comment by tivervac** [comment]
@Rainer7000 We might be seeing similar issues. What does `not rendering` mean? In our case, events are still being processed (moving windows, copying and pasting...) but nothing is drawn. Making the window smaller than larger again just causes the resized parts to go full black

**Comment by Rainer7000** [comment]
@tivervac Yes, I think this is the same behaviour as in our case. The content of the editor tab (Composites, Labels etc.) gets created but nothing is drawn. The tab in the third image in my original post stays gray but should look like the tab in the first image with a left and right label on white background.

**Comment by Rainer7000** [comment]
After coloring my test-composites (left=red, right=green) I found out that after closing the second window there is a area with the color of the window background drawn in front of everything. You can see this if you enlarge the first window after the effect occours. It seems to have the size of the editor-content when the second window is closed.
I hope this helps to find the bug.
![eclipse_2023_09_problem](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/4f485154-fd6a-4e5b-8f8b-92c337cce9d3)

**Comment by iloveeclipse** [comment]
> The Problem was introduced with Eclipse 2023-09, alle previous versions workt fine.
> I hope this helps to find the bug.

Could you try to bisect the issue?

**Comment by Rainer7000** [comment]
I'm not shure what you mean by 'bisect'? There is only one issue.
The only way to trigger the problem is closing a second workbench window while the editor-earea of the first window is empty. 
After that all newly opened editors will have a background square drawn over the content area. 
I could try to zip my test-project and attach it here if that would help.

**Comment by iloveeclipse** [comment]
> I'm not sure what you mean by 'bisect'? 

I meant "git bisect" - common approach used to find out faulty commit.

> I could try to zip my test-project and attach it here if that would help.

This would help for sure, but I hoped you could point to the commit in question. I personally have no time to bisect, especially if to see the regression many manual steps are needed. But may be someone affected by this issue could use your project to provide that info.

Once "bad" commit is identified it would be much easier to understand what is broken and how to fix that.

**Comment by Rainer7000** [comment]
As I have neither enough experience in git (other than the mere basics) nor with github or the eclipse core mechanisms I can't investigate the issue further. My assumtion would be, that it has something to do with the new background image that can be drawn in the empty editor area wich was introduced in 2023-09.
![editor_background](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/95d56ba6-4225-4e9f-b3e8-1f4524b2d4a3)

**Comment by Rainer7000** [comment]
I attach here the simple Test-App to reproduce the Problem
[zzUITestApp.zip](https://github.com/eclipse-platform/eclipse.platform.ui/files/12816775/zzUITestApp.zip)

**Comment by Rainer7000** [comment]
After quite some time I had the opportunity to investigate this bug further. It turned out you don't need the the test-application above and it has nothing to do with using the eclipse 3 Workbench-API. 
You can trigger the bug in a plain eclipse 2023-09.
Here ist the steps to reproduce it:
- Open up eclipse and make your Editor-Area small. Also close alle open Editors.
- From the menu Select "Window" -> "New Window" to open a second Workbench Window.
- Switch to the initial Window, open a Editor and close it again so that you can see the "Onboarding"-Image and Text
- Now Close the the second Window.
- If you now open the Editor from before you still see the Onboarding-Area. 
    -> If you enlarge the Editor-Area you can see that the Editor-Content is covered by the Onboarding-Widgets.

![onboarding](https://github.com/eclipse-platform/eclipse.platform.ui/assets/57557879/09e8a52c-7a10-4309-910e-b0f469d2eae3)

**Comment by BeckerWdf** [comment]
@marcushoepfner: Do you have any clue what's going on here?

**Comment by marcushoepfner** [comment]
The onboarding  image and the onboarding label are overwritten when the second workbench window is opened. On close they are disposed. That's why the first window doesn't do any processing of the onboarding stuff.

The solution is, to prevent any state in StackRenderer. We must not store onboardingComposite, onboardingLabel and onboardingImage in fields of stackrender. Instead we should retrieve them from tabfolder. I'm working on a solution.
```java
	private void createOnboardingControls(CTabFolder tabFolder) {
		Composite onBoarding = WidgetFactory.composite(SWT.NONE).layout(GridLayoutFactory.swtDefaults().create())
				.background(tabFolder.getBackground()).create(tabFolder);

		onboardingComposite = WidgetFactory.composite(SWT.NONE).background(tabFolder.getBackground())
				.create(onBoarding);
		GridDataFactory.create(GridData.VERTICAL_ALIGN_CENTER | GridData.HORIZONTAL_ALIGN_CENTER).grab(true, true)
				.applyTo(onboardingComposite);

		GridLayoutFactory.swtDefaults().numColumns(2).equalWidth(true).spacing(10, SWT.DEFAULT)
				.applyTo(onboardingComposite);

		GridDataFactory gridDataFactory = GridDataFactory.swtDefaults().align(SWT.CENTER, SWT.CENTER)
				.indent(SWT.DEFAULT, 10).span(2, 1);

		onboardingImage = WidgetFactory.label(SWT.NONE).supplyLayoutData(gridDataFactory::create)
				.create(onboardingComposite);

		Color color = JFaceResources.getColorRegistry().get(JFacePreferences.QUALIFIER_COLOR);
		onboardingText = WidgetFactory.label(SWT.NONE).foreground(color).supplyLayoutData(gridDataFactory::create)
				.create(onboardingComposite);

		onBoarding.setLocation(ONBOARDING_SPACING, ONBOARDING_TOP_SPACING);
		Consumer<ControlEvent> sizeUpdate = e -> setOnboardingControlSize(tabFolder, onBoarding, onboardingImage);
		tabFolder.addControlListener(ControlListener.controlResizedAdapter(sizeUpdate));
		Consumer<CTabFolderEvent> tabCountUpdate = e -> setOnboardingControlSize(tabFolder, onBoarding,
				onboardingImage);
		tabFolder.addCTabFolder2Listener(CTabFolder2Listener.itemsCountAdapter(tabCountUpdate));

		tabFolder.addDisposeListener(e -> {
			if (onboardingImage != null && !onboardingImage.isDisposed() && onboardingImage.getImage() != null
					&& !onboardingImage.getImage().isDisposed()) {
				onboardingImage.dispose();
			}
		});
	}
```

**Comment by marcushoepfner** [comment]
I would need some help w.r.t. to the E4 model.

How can I get the editor CTabFolder from the MPerspective element which I get when the perspective is changed? I mean in org.eclipse.e4.ui.workbench.renderers.swt.StackRenderer.
I found methods like findItem or using the modelServervice but I don't know how to use them. @vogella, can you help here?

Coding looks like this:
```java
	@Inject
	@Optional
	void subscribePerspectiveSwitched(
			@UIEventTopic(UIEvents.UILifeCycle.PERSPECTIVE_SWITCHED) org.osgi.service.event.Event event) {
		Object element = event.getProperty(EventTags.ELEMENT);
		if (element instanceof MPerspective) {
			setOnboarding((MPerspective) element);
		}
	}

	private void setOnboarding(MPerspective perspective) {
		//TODO: get editor stack CTabFolder from perspective
		//...
		
		if (onboardingImage == null || onboardingText == null || onboardingComposite == null
				|| onboardingComposite.isDisposed()) {
			return;
		}

		Arrays.stream(onboardingComposite.getChildren()).filter(c -> c != onboardingImage)
				.filter(c -> c != onboardingText).forEach(Control::dispose);

		String textId = "persp.editorOnboardingText:"; //$NON-NLS-1$
		perspective.getTags().stream().filter(tag -> tag.startsWith(textId)).map(tag -> tag.substring(textId.length()))
				.findFirst().ifPresentOrElse(onboardingText::setText, () -> onboardingText.setText("")); //$NON-NLS-1$

		String iconId = "persp.editorOnboardingImageUri:"; //$NON-NLS-1$
		perspective.getTags().stream().filter(tag -> tag.startsWith(iconId)).map(tag -> tag.substring(iconId.length()))
				.findFirst().map(this::getImageFromURI).filter(Predicate.not(Image::isDisposed))
				.ifPresentOrElse(onboardingImage::setImage, () -> onboardingImage.setImage(null));

		String commandId = "persp.editorOnboardingCommand:"; //$NON-NLS-1$
		String[] commands = perspective.getTags().stream().filter(tag -> tag.startsWith(commandId))
				.map(tag -> tag.substring(commandId.length())).toArray(String[]::new);

		Color color = JFaceResources.getColorRegistry().get(JFacePreferences.QUALIFIER_COLOR);

		GridDataFactory labelGridDataFactory = GridDataFactory.swtDefaults().align(SWT.RIGHT, SWT.CENTER);
		GridDataFactory commandGridDataFactory = GridDataFactory.swtDefaults().align(SWT.LEFT, SWT.CENTER);

		for (int i = 0; i < commands.length; i++) {
			labelGridDataFactory.indent(SWT.DEFAULT, i == 0 ? 10 : SWT.DEFAULT);
			commandGridDataFactory.indent(SWT.DEFAULT, i == 0 ? 10 : SWT.DEFAULT);

			String[] commandAndText = commands[i].split("\\$\\$\\$"); //$NON-NLS-1$

			WidgetFactory.label(SWT.NONE).text(commandAndText[0]).foreground(color)
					.supplyLayoutData(labelGridDataFactory::create).create(onboardingComposite);
			WidgetFactory.label(SWT.NONE).text(commandAndText[1]).foreground(color)
					.supplyLayoutData(commandGridDataFactory::create).create(onboardingComposite);
		}
	}
```

**Comment by vogella** [comment]
So you have the MPerspective and the the EModelService and would like to the MArea?

**Comment by laeubi** [comment]
You need to call `getObject` on the model to get the underlying SWT control(s).

**Comment by marcushoepfner** [comment]
> So you have the MPerspective and the the EModelService and would like to the MArea?

Not sure what I want :)

Yes, MPerspective and EModelService and I need to get the CTabFolder for the editor.
If I see correctly it has been created from an MPartStack with tag "EditorStack" (method createWidget in StackRenderer). 
So I guess I need the MPartStack

**Comment by vogella** [comment]
Did you check via the model spy what you need? Currently not at the computer but you can open it via Windows -> Spies. From your perspective search downwards for the element you want and you can find it via modelService.findElements(parent, Id (or null, Modelelemet.class (or null) IIRC

**Comment by iloveeclipse** [comment]
@marcushoepfner : is this bug fixed now with #1400, or is something left to do? Please close if fixed.

**Comment by marcushoepfner** [comment]
sorry, always forget this.
